### PR TITLE
Enable torchdynamo with torch_tensorrt(fx path)

### PR DESCRIPTION
**Pull request by frank-wei**
# What does this PR do?
 Adding support for TorchDynamo with torch_tensor(fx2trt module). Detailed context available at #17724 
This diff is about adding extra inference backend based on #17308 
 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## To reproduce and set up the environment
```
# install torch-nightly
conda install pytorch torchvision torchaudio cudatoolkit=11.3 -c pytorch-nightly

# install functorch (and reinstall after `git pull` later if need to sync up)
git clone https://github.com/pytorch/functorch
cd functorch
rm -rf build
pip install -e .[aot]

cd ..
git clone https://github.com/pytorch/torchdynamo
cd torchdynamo
pip install -r requirements.txt
python setup.py develop

# install TensorRT
pip install nvidia-pyindex
pip install nvidia-tensorrt==8.2.4.2

# install torch_tensorrt (fx path)
cd ..
git clone https://github.com/pytorch/TensorRT.git
cd TensorRT/py
python setup.py install --fx-only

```
cc HF @stas00 
cc Meta @yinghai @Chillee 
cc NV @ncomly-nvidia @narendasan

### CODE REVIEW

**Comment by frank-wei** [comment]
Hi, @stas00  just a friendly ping. I updated the installation part and it will be easy to repro if needed.

**Review by stas00 — CHANGES_REQUESTED** [review]
I was away, catching up now.

trying to figure out how to install TensorRT - so can't yet appreciate your PR.  Meanwhile there is already work to do - please see comments.

Additionally the installation info you shared needs to be part of the docs, since otherwise the users won't know what to do. Of course, ideally it should be on your side and HF docs link to it. As that way you can easily keep it up-to-date. Thank you.

And of course the new API options need to be documented as well in the user-facing documentation otherwise nobody will use this feature. @anijain2305, could you please complete that long overdue documentation? We won't be able to finish this PR until the doc task from https://github.com/huggingface/transformers/pull/17308#issuecomment-1157239692 is completed. Thank you! You can of course do it as part of this PR and document the overdue options and the new ones in one go.

**Comment by frank-wei** [comment]
Thanks for your time and efforts @stas00 !
1. Yeah, the TRT seems that bring the new user some troubles when they try their first time to install. I just found a way to install python version of TRT so you do not need to download TRT tarball and unzip the stuffs (this python installation will install all the dependent libs like tensorRT lib and cuDNN lib). I added this instructions to our doc as a PR. https://github.com/pytorch/TensorRT/pull/1145/
```
    $ pip3 install nvidia-pyindex
    $ pip3 install nvidia-tensorrt==8.2.4.2
```
2. I am having a PR to **disable** the bazel check https://github.com/pytorch/TensorRT/pull/1147. Will update once it is merged.
But that is a bit weird for bazel installation. I am on centOS and conda envrioment. Here is command `conda install -c conda-forge bazel`. It looks like your bazel installation location is not added to $PATH but `which bazel` can help check. Anyway, with my diff [1147](https://github.com/pytorch/TensorRT/pull/1147), we should not need bazel. And below instructions is the complete one I just verified work.
```
    $ conda create --name python_env python=3.8
    $ conda activate python_env
    # Recommend to install PyTorch 1.12 and later
    $ conda install pytorch torchvision torchtext cudatoolkit=11.3 -c pytorch-nightly
    # Install TensorRT python package
    $ pip3 install nvidia-pyindex
    $ pip3 install nvidia-tensorrt==8.2.4.2
    $ git clone https://github.com/pytorch/TensorRT.git
    $ cd TensorRT/py && python setup.py install --fx-only && cd ..
    # check torch_tensorrt.fx is installed
    $ python -c "import torch_tensorrt.fx"
  ```
  Hope it solves your problem.

File: tests/trainer/test_trainer.py
Comment (frank-wei) [reply tests/trainer/test_trainer.py:1751]: 1. The test passed because we set the rule in TorchDynamo as if the TRT lowering by torch_tensorrt.fx is failed, the model will fall back eager mode. So I think that is reason the test still works.
2. Good question
> which of course begs the question - where in the tests do you check if torch_tensorrt is available?

I will add a `try` block to check if `import torch_tensorrt.fx` is correctly returned.

**Comment by stas00** [comment]
`conda install -c conda-forge bazel` did the trick same with pip was giving nothing with `which bazel` - not a PATH issue, but a package issue I think.

---------------

```
    $ pip3 install nvidia-pyindex
    $ pip3 install nvidia-tensorrt==8.2.4.2
```

That did the trick. The tests have run successfully.

so let's update the OP with the above 2 fixes.

File: tests/trainer/test_trainer.py
Comment (stas00) [reply tests/trainer/test_trainer.py:1751]: Yeah, let's figure out how to not have an assert and a passing test at the same time ;) Thank you!

Comment (stas00) [reply tests/trainer/test_trainer.py:1751]: but also if the user failed to install tensor RT properly (like I did) and now the API hides that - how will the user know that their setup is incomplete/broken?

**Comment by stas00** [comment]
ah, one more user-facing documentation - if you want users to use your magic code you will need to provide some enticement. A small benchmark usually goes a long way to get a user excited to try a new feature. So this is something else to consider. It's not a show stopper, but as you can see if the docs aren't added right away they never get added, so it's best to do it in one go. It's still a recommendation and I'm fine merging it as is, it's just not going to be used much w/o enticing docs.

File: tests/trainer/test_trainer.py
Comment (frank-wei) [reply tests/trainer/test_trainer.py:1751]: Actually, similar to other library like pytorch, users will be alerted by
1) in our installation, we asked user to try `python -c "import torch_tensorrt.fx"` or they will be alerted by stdout by
ModuleNotFoundError: No module named 'torch_tensorrt'
2) if fx lowering is failed, a print statement is there like 
https://github.com/pytorch/torchdynamo/blob/5fb502660e52a2e1f93ab0f148fd8776e1b56297/torchdynamo/optimizations/backends.py#L742

Comment (frank-wei) [reply tests/trainer/test_trainer.py:1751]: one question, is this `test_trainer` in CI/CD? Should we add torch_tensorrt.fx as one dependency of environment setup? 
Or even have `@require_torchtensorrt.fx` ?

**Comment by frank-wei** [comment]
> ah, one more user-facing documentation nit - if you want users to use your magic code you will want to provide some enticement. A small benchmark table that shows what these features do usually goes a long way to get a user excited to try them. So this is something else to consider. It's not a show stopper, but as you can see if the docs aren't added right away they never get added, so it's best to do it in one go. It's still a recommendation and I'm fine merging it as is, it's just not going to be used much w/o enticing docs.

I will try to add the doc there. But it is better to have @anijain2305  to include the AOT part.:-)

**Comment by stas00** [comment]
> I will try to add the doc there. But it is better to have @anijain2305 to include the AOT part.:-)

Yeah, I was hoping that you'd only need to add the incremental part relevant for this PR.

File: tests/trainer/test_trainer.py
Comment (stas00) [reply tests/trainer/test_trainer.py:1751]: re: diagnostics - please let's save all your notes in the docs -> or if they are already in your docs then direct links please (i.e. "here is how to diagnose things")

**Comment by stas00** [comment]
re: CI - yes and it's complicated

basically the live CI that you see reporting in this PR runs only CPU tests since CircleCI doesn't have gpus.

then we have another set of CI workflows that runs on our machine via github actions and that's where we test all the complex/slow cases.

And yes, I completely forgot that part of this PR we need to setup our CI to install all these packages as well so that these tests will be run.

So once we polished this let's not forget that part. We will have to repeat all those instructions - but actually there is a problem - how will the docker builder be able to download tensorRT packages if they require a user account?

File: tests/trainer/test_trainer.py
Comment (frank-wei) [reply tests/trainer/test_trainer.py:1751]: Will add the user self-diagnostics to docs.

**Comment by stas00** [comment]
> Actually, circleCI has gpu resource to use(V100, T4, P4). I just added to our project :-) [pytorch/TensorRT#1137](https://github.com/pytorch/TensorRT/pull/1137) 

That's great to know - thank you very much - I will pass this info on

> These 2 commands are our saver
> 
> ```
>  $ pip3 install nvidia-pyindex
>  $ pip3 install nvidia-tensorrt==8.2.4.2
> ```

Ah, right! so no need for nvidia user account! super - let's use that in the instructions then.

> Do you think we need to have @require_torchtensorrt.fx ? So it will help us to check if torch_tensorrt.fx is installed in the test?

Absolutely, yes!

**Comment by frank-wei** [comment]
@stas00 , just wondering if the circleci is flaky? Some tests errors are not related. For ex.
run_example_torch, check_code_quanlity

**Comment by stas00** [comment]
It appears that the CI is very broken at the moment, I asked and will know more tomorrow morning.

Thank you for the heads up, @frank-wei - it doesn't look like any of the failures are related to your work. Especially since the live CI won't run any of your tests.

**Comment by stas00** [comment]
ok, so for the quality one - please rebase this PR on main. Thank you.

The other issue I don't have an answer for yet.

**Comment by stas00** [comment]
ok, so to fix `check_code_quality` you need to run `make style` and push

but the rest of the problems are now coming from this PR:

```

==================================== ERRORS ====================================
______________ ERROR collecting tests/deepspeed/test_deepspeed.py ______________
ImportError while importing test module '/home/circleci/transformers/tests/deepspeed/test_deepspeed.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/local/lib/python3.7/importlib/__init__.py:127: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/deepspeed/test_deepspeed.py:26: in <module>
    from tests.trainer.test_trainer import TrainerIntegrationCommon  # noqa
tests/trainer/test_trainer.py:586: in <module>
    class TrainerIntegrationTest(TestCasePlus, TrainerIntegrationCommon):
tests/trainer/test_trainer.py:1803: in TrainerIntegrationTest
    @require_torch_tensorrt_fx
src/transformers/testing_utils.py:499: in require_torch_tensorrt_fx
    return unittest.skipUnless(is_torch_tensorrt_fx_available(), "test requires Torch-TensorRT FX")(test_case)
src/transformers/utils/import_utils.py:421: in is_torch_tensorrt_fx_available
    return importlib.util.find_spec("torch_tensorrt.fx") is not None
/usr/local/lib/python3.7/importlib/util.py:94: in find_spec
    parent = __import__(parent_name, fromlist=['__path__'])
E   ModuleNotFoundError: No module named 'torch_tensorrt'
```

Let me know if you need help with sorting it out.

**Comment by frank-wei** [comment]
Thanks @stas00 , I fixed the import check.

File: docs/source/en/perf_train_gpu_one.mdx
Comment (stas00) [inline docs/source/en/perf_train_gpu_one.mdx:722]: > can be deployed to lower to an optimized engine

`it` is missing? lower what - `it` being `the fx graph`? did I guess it right? perhaps you meant something else

File: src/transformers/trainer.py
Comment (stas00) [inline src/transformers/trainer.py:2295]: I'm not sure this is correct - shouldn't torchdynamo's first 2 options work w/o needing tensorRT?

If so then this needs to broken up and tensorrt check moved inside only for the 2 options relevant to it.

Comment (stas00) [reply src/transformers/trainer.py:2295]: Also I think the original code has an issue here and the new code copies it.

what happens if either is not available and the user has no idea their `--torchdynamo` option is being silently ignored. Unless this is done somewhere else but I don't see it.

I think inside `training_args.py` - `torchdynamo` should check that it's available if the option is set and assert if it's not. Moreover there is no need to slow down the context manager with a huge if/else tree. Why not figure that out at argparse stage (`training_args.py` - set a custom final ctx_manager and assign if it's not None.

Let me know if you'd like to re-work it, @frank-wei - I don't want to impose it on you since this code predates your contributions.

File: docs/source/en/perf_train_gpu_one.mdx
Comment (frank-wei) [reply docs/source/en/perf_train_gpu_one.mdx:722]: Yes, you are right. I will fix it.

File: src/transformers/trainer.py
Comment (frank-wei) [reply src/transformers/trainer.py:2295]: >I think inside training_args.py - torchdynamo should check that it's available if the option is set and assert if it's not.

I agree. It should be able to assert at very early stage.

>Why not figure that out at argparse stage (training_args.py - set a custom final ctx_manager and assign if it's not None.

I can help to move the ctx manager to arg stage.

**Comment by pommedeterresautee** [comment]
Is there a simple way to support dynamic shape on fx2trt on torch Dynamo? If not yet, may be you want to specify it in the doc? If yes you may want to say how we provide the Tensorrt "profiles" ?

In rapid experiments I did on HF + dynamo + fx2trt, even by increasing the dynamo cache, when I pushed plenty of different input sizes, at some point it raised plenty of OOM exceptions and stopped working. May be trt profiles would have worked.

**Comment by frank-wei** [comment]
> Is there a simple way to support dynamic shape on fx2trt on torch Dynamo? If not yet, may be you want to specify it in the doc? If yes you may want to say how we provide the Tensorrt "profiles" ?
> 
> In rapid experiments I did on HF + dynamo + fx2trt, even by increasing the dynamo cache, when I pushed plenty of different input sizes, at some point it raised plenty of OOM exceptions and stopped working. May be trt profiles would have worked.

It is not supported yet for dynamic shape in my implementation. I plan to support dynamic batch size in the next step(probably set it to default).

File: src/transformers/trainer.py
Comment (stas00) [reply src/transformers/trainer.py]: please ping me when you're ready for another review. Thank you, @frank-wei!

Comment (frank-wei) [reply src/transformers/trainer.py]: Hi, @stas00 , the changes are pushed 3 days ago. I moved the ctx manager to training_args.py. Please let me know if you can not find it.

Comment (stas00) [reply src/transformers/trainer.py]: yes, but you have been pushing multiple changes, so I'm just waiting for you to give me a green light to review the PR again, as I have no idea when you're done. Does it make sense?

i.e. in the future please signal us when we need to look again.

Comment (frank-wei) [reply src/transformers/trainer.py]: > please ping me when you're ready for another review. Thank you, @frank-wei!

I see. Sounds good to me. Now it is a good time to review again. Thanks, @stas00

**Comment by stas00** [comment]
One of the tests is failing for me:

```
$ CUDA_VISIBLE_DEVICES=0 RUN_SLOW=1 pyt tests/trainer/test_trainer.py -k test_torchdynamo_memory -sv
        # AOT Autograd recomputaion and nvfuser recomputation optimization
        # aggressively fuses the operations and reduce the memory footprint.
>       self.assertGreater(orig_peak_mem, peak_mem * 2)
E       AssertionError: 100664832 not greater than 201330688
```

let me know what details you need - this is on A100.

File: src/transformers/training_args.py
Comment (stas00) [inline src/transformers/training_args.py:1246]:
```suggestion
        if self.torchdynamo:
            if not is_torchdynamo_available():
                raise RuntimeError("Torchdynamo is not installed.")

            import torchdynamo
            from torchdynamo.optimizations import backends
            from torchdynamo.optimizations.training import aot_autograd_speedup_strategy

            def get_ctx():
                # Normal
                if self.torchdynamo == "eager":
                    return torchdynamo.optimize("eager")
                elif self.torchdynamo == "nvfuser":
                    return torchdynamo.optimize(aot_autograd_speedup_strategy)

                # TensorRT
                if self.torchdynamo in ["fx2trt-fp16", "fx2trt"]:
                    if not is_torch_tensorrt_fx_available():
                        raise RuntimeError("Torch-TensorRT FX path is not installed.")

                    if self.torchdynamo == "fx2trt-fp16":
                        return torchdynamo.optimize(backends.fx2trt_compiler_fp16)
                    elif self.torchdynamo == "fx2trt":
                        return torchdynamo.optimize(backends.fx2trt_compiler)

                raise RuntimeError(f"Torchdynamo backend {self.torchdynamo} is not supported.")
            
            self.ctx_manager_torchdynamo = get_ctx()
        else:
            self.ctx_manager_torchdynamo = contextlib.nullcontext()
```

I've tried to make the very long ifelse a little bit easier to read and removing repetition. let me know if this resonates.

Untested.

Comment (stas00) [reply src/transformers/training_args.py:1246]: and if the number of options continues to grow, since it's quite symmetric we can just make a dict and look it up for the context and the requirements. But this is probably good enough for now.

**Comment by frank-wei** [comment]
The failure may due to the torchdynamo outdated? Could you install the newest torchdynamo? Here are the command to install it:
```
git clone https://github.com/pytorch/functorch
cd functorch
rm -rf build
pip install -e .[aot]

cd ..
git clone https://github.com/pytorch/torchdynamo
cd torchdynamo
pip install -r requirements.txt
python setup.py develop
```

It looks good from my testing:
```
(mypy38-fx-only) [wwei6@devgpu005.ftw6 /data/users/wwei6/Work/transformers] CUDA_VISIBLE_DEVICES=6 pytest tests/trainer/test_trainer.py  -k test_torchdynamo_memory -sv
===================================================================================== test session starts =====================================================================================
platform linux -- Python 3.8.13, pytest-7.1.2, pluggy-1.0.0 -- /data/users/wwei6/miniconda3/envs/mypy38-fx-only/bin/python
cachedir: .pytest_cache
benchmark: 3.4.1 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
hypothesis profile 'default' -> database=DirectoryBasedExampleDatabase('/data/users/wwei6/Work/transformers/.hypothesis/examples')
rootdir: /data/users/wwei6/Work/transformers, configfile: setup.cfg
plugins: benchmark-3.4.1, hydra-core-1.1.2, hypothesis-6.49.1
collected 70 items / 69 deselected / 1 selected                                                                                                                                               

tests/trainer/test_trainer.py::TrainerIntegrationTest::test_torchdynamo_memory PyTorch: setting up devices
The default value for the training argument `--report_to` will change in v5 (from all installed integrations to none). In v5, you will need to use `--report_to all` to get the same behavior as now. You should start updating your code and make this info disappear :-).
PASSED

====================================================================================== warnings summary =======================================================================================
../../miniconda3/envs/mypy38-fx-only/lib/python3.8/site-packages/torch/utils/tensorboard/__init__.py:4
  /data/users/wwei6/miniconda3/envs/mypy38-fx-only/lib/python3.8/site-packages/torch/utils/tensorboard/__init__.py:4: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    if not hasattr(tensorboard, "__version__") or LooseVersion(

../../miniconda3/envs/mypy38-fx-only/lib/python3.8/site-packages/torch/utils/tensorboard/__init__.py:6
  /data/users/wwei6/miniconda3/envs/mypy38-fx-only/lib/python3.8/site-packages/torch/utils/tensorboard/__init__.py:6: DeprecationWarning: distutils Version classes are deprecated. Use packaging.version instead.
    ) < LooseVersion("1.15"):

tests/trainer/test_trainer.py::TrainerIntegrationTest::test_torchdynamo_memory
  /data/users/wwei6/miniconda3/envs/mypy38-fx-only/lib/python3.8/site-packages/torch/nn/utils/_stateless.py:5: DeprecationWarning: The `torch.nn.utils._stateless` code is deprecated now that it is publicly available. Please use `torch.nn.utils.stateless instead.
    warnings.warn("The `torch.nn.utils._stateless` code is deprecated now that "

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
======================================================================== 1 passed, 69 deselected, 3 warnings in 7.51s =========================================================================
```

File: src/transformers/training_args.py
Comment (frank-wei) [reply src/transformers/training_args.py]: Thanks for the efforts of refactoring. I took it in the change.

**Comment by stas00** [comment]
I suspected that was the case, but what I was trying to say is that the test should have failed on the torchdynamo error and not the mismatch in values, i.e. something is trapping the real error and the user could be not the wiser that their torchdynamo is broken - e.g. when there are a lot of logs.

It needs to assert on the actual error. Does it make sense?

**Comment by frank-wei** [comment]
> I suspected that was the case, but what I was trying to say is that the test should have failed on the torchdynamo error and not the mismatch in values, i.e. something is trapping the real error and the user could be not the wiser that their torchdynamo is broken - e.g. when there are a lot of logs.
> 
> It needs to assert on the actual error. Does it make sense?

hm.. that is something out of my expertise as it relates with torchdynamo. If it is torch_tensorrt related, I'd love to help.

For the CI test error, it seems that test is flaky? I did not find useful information. Could you help triage that? Thanks.

File: tests/trainer/test_trainer.py
Comment (stas00) [inline tests/trainer/test_trainer.py]: I removed this requirement, since it's not being used. please correct me if I'm wrong.

**Review by stas00 — APPROVED** [review]
Hi @frank-wei, I had to rebuild the whole environment against pt-nightly and now everything works.

I think it'd be good to save the instructions in the OP somewhere so that it's easier for the user and us to be able to rebuild the environment.

Would you like to maintain a section or a file on your side that contains the instructions in the OP and we could point to it?

Other than that,  I will just ask Sylvain to have a quick review and we can merge this.

Thank you for your patience.

File: tests/trainer/test_trainer.py
Comment (frank-wei) [reply tests/trainer/test_trainer.py]: Thanks for the catching! This test does not use tensorrt.

**Comment by frank-wei** [comment]
> Hi @frank-wei, I had to rebuild the whole environment against pt-nightly and now everything works.
> 
> I think it'd be good to save the instructions in the OP somewhere so that it's easier for the user and us to be able to rebuild the environment.
> 
> Would you like to maintain a section or a file on your side that contains the instructions in the OP and we could point to it?
> 
> Other than that, I will just ask Sylvain to have a quick review and we can merge this.
> 
> Thank you for your patience.



> Hi @frank-wei, I had to rebuild the whole environment against pt-nightly and now everything works.
> 
> I think it'd be good to save the instructions in the OP somewhere so that it's easier for the user and us to be able to rebuild the environment.
> 
> Would you like to maintain a section or a file on your side that contains the instructions in the OP and we could point to it?
> 
> Other than that, I will just ask Sylvain to have a quick review and we can merge this.
> 
> Thank you for your patience.

Thanks @stas00 , do you think I can add a 3 pointers for installations of torchdynamo, functorch, torch_tensorrt in docs/source/en/perf_train_gpu_one.mdx ?

**Comment by stas00** [comment]
I think that works, @frank-wei

**Review by sgugger — COMMENTED** [review]
Thanks for your PR! I left some comment on where the context manager "getter" should live, as it shouldn't be an attribute of `TrainingArguments`, a dataclass we serialize at each save.

File: src/transformers/training_args.py
Comment (frank-wei) [reply src/transformers/training_args.py]: @sgugger, that sounds good to me. I moved the function to `integrations.py`

Comment (stas00) [reply src/transformers/training_args.py]: @sgugger, the original implementation `calculated` the context on every call - that's why I suggested to move the logic to the argrparse stage, since this logic needs to be done only once.

What would be a good place then to perform this figuring out? In trainer's init probably, right?

Comment (sgugger) [reply src/transformers/training_args.py]: That works, yes.

Comment (stas00) [reply src/transformers/training_args.py]: @frank-wei, please let me know if you need help here - moving to trainer's init that is.

**Comment by frank-wei** [comment]
@stas00  @sgugger  please check the change. The failed test seems flaky and not related.

**Review by sgugger — APPROVED** [review]
Way better this way, thanks for iterating! Left a last nit.

File: src/transformers/trainer.py
Comment (frank-wei) [reply src/transformers/trainer.py]: Cool. Fixed that.

Comment (sgugger) [reply src/transformers/trainer.py]: Thanks!

**Comment by sgugger** [comment]
@stas00 Are you good with this last iteration (as long as all tests pass?)

**Comment by stas00** [comment]
Let me run the tests.

**Comment by stas00** [comment]
All tests pass.

**Pull request merged**
